### DISCUSSION TITLE
C-x 1 --> 3 : How do I maintain split window proportions?

### DISCUSSION THREAD

**Discussion by aykaramba** [lem-project/lem#1266] [General]
Hi everyone.

LEM is my day to day driver and it makes the editor experience so nice.  I am running into an issue that is becoming difficult to overlook.  I split my LEM window into various configurations for different projects and workflows and the issue that I am running into is that LEM will automatically resize the frames.

For any project, I am loathe to actually ask for anything at all because the gift of the code is merely enough to humble me an dI just power through it.  This issue, however, is becoming difficult to ignore as it is really impacting my wofklow on a particular project.

I know there are opinions about how this should work and that the behaviour of resizing all window splits to a balanced layout is a feature and not a bug.  However, I am looking for a quick hack that would allow me stop LEM from resizing my window splits until some new approach is developed.

Does anyone know if it is possible to give LEM a variable value to not perform this behaviour or maybe there is a way to change a function to accomplish the same?

Any help is greatly appreciated.

**Comment by cxxxr** [comment]
Hi, thanks for using lem.
This is something I would like to improve too, but haven't been able to.
I guess a work-around would be to set `lem-core/commands/window::*balance-after-split-window*` to nil.

**Comment by aykaramba** [comment]
Hi! Thank you kindly for this amazing editor.

I set the var to nil as follows: (defvar *balance-after-split-window* nil)

And then I recompiled and ran the sdl2 binary.  Additional behaviour details:

1) I have to do some more testing but as long as the window is focused and I keep on working in the editor the window splits don't seem to resize.  I need to use it some more to confirm.

2) However, whenever i focus another window and then re-focus the Lem window the window slices are rebalanced. 

Is there another *balance-after-split-window* variable that can be changed to prevent #2 above?

Thanks.

**Comment by cxxxr** [comment]
I think defvar is not reflected if the variable is already bound
What about using setf/setq?

**Comment by aykaramba** [comment]
Okay, here is (perhaps) some interesting information:

1) setq / setf did not work.

2) However,  I hacked up the balance-windows function in the window-tree.lisp file to not balance windows as follows:

``` 

(defun balance-windows (&optional (window-tree (window-tree)))
  (labels ((rec
	     (node x y width height)
             (cond ((windowp node)
                    (window-set-size node width height)
                    (window-set-pos node x y))


                    (t

                       ;; (destructuring-bind (split-type &rest children) node
		       ;; 			(let ((n (length children)))
		       ;; 			  (ecase split-type
		       ;; 				 (:hsplit
		       ;; 				  (multiple-value-bind (child-width remaining) (floor width n)
		       ;; 						       (loop :with rest-width := width
		       ;; 							     :for (child . rest-children) :on children
		       ;; 							     :for first := t :then nil
		       ;; 							     :do (let ((child-width child-width))
		       ;; 								   (cond ((null rest-children)
		       ;; 									  (setf child-width rest-width))
                       ;; 									 ((plusp remaining)
		       ;; 									  (decf remaining)
		       ;; 									  (incf child-width)))
		       ;; 								   (rec child
		       ;; 									x
		       ;; 									y
		       ;; 									(if (null rest-children)
		       ;; 									    child-width
		       ;; 									  (- child-width
		       ;; 									     (frame-window-left-margin (current-frame))))
		       ;; 									height)
		       ;; 								   (incf x child-width)
		       ;; 								   (decf rest-width child-width)))))
		       ;; 				 (:vsplit
		       ;; 				  (multiple-value-bind (child-height rem) (floor height n)
		       ;; 						       (loop :with rest-height := height
		       ;; 							     :for (child . rest-children) :on children
		       ;; 							     :do (let ((child-height child-height))
		       ;; 								   (cond ((null rest-children)
		       ;; 									  (setf child-height rest-height))
		       ;; 									 ((plusp rem)
		       ;; 									  (decf rem)
		       ;; 									  (incf child-height)))
		       ;; 								   (rec child
		       ;; 									x
		       ;; 									y
		       ;; 									width
		       ;; 									child-height)
		       ;; 								   (incf y child-height)
		       ;; 								   (decf rest-height child-height))))))))

		     nil
		     
		     ) 



		   )))


	  (rec (convert-window-n-tree window-tree)
               (topleft-window-x (current-frame))
               (topleft-window-y (current-frame))
               (max-window-width (current-frame))
               (max-window-height (current-frame)))))


```

3) I am learning Lisp and I can read enough to understand what should be going on to turn off functionality.  I hope this hack is okay, for now it seems to work.  I hope to learn enough Lisp to start helping out, this project is great and the code is very readable even for a noob.

4) Cxxxr, please do not let my question distract from anything you are working on.  My balance-windows issue is not a big deal, this more than solves the problem if it doesn't cause an issue anywhere else.

Thanks!

**Reply by aykaramba** [reply]
Note to self:  The file is located here: lem/src/window/window-tree.lisp

**Comment by aykaramba** [comment]
Just dropping in to say this works in 2.3.0 as well.  Note, this breaks the "reloading" of split frames when you resize fonts.  To work around this you have to C-x 0 to a single frame, size the fonts down and then split up the frames again.  I need to provide a patch for this at some point.

Thanks for the great hackable editor!
